Added Zally validation to extension

semonte opened the pull request:
This commit enables Zally validation. A user must
configure a Zally URL where the YAML spec is sent.

A user must also have a `ztoken` command available in
order for the plugin to create authenticated requests
to the Zally API.

semonte committed:
Added Zally validation to extension

This commit enables Zally validation. A user must
configure a Zally URL where the YAML spec is sent.

A user must also have a `ztoken` command available in
order for the plugin to create authenticated requests
to the Zally API.

maxim-tschumak commented on examples/extensions-zalando/src/main/java/org/zalando/intellij/swagger/examples/extensions/zalando/validator/zally/YamlFileValidator.java:
```diff
@@ -0,0 +1,89 @@
+
+    private LintingResponse validate(final PsiFile file) {
+        try {
+            final ZallyService zallyService = ServiceManager.getService(ZallyService.class);
```
You create a Zally service every time the `validate` function is called. Let's do it once, during the initialization.

semonte replied:
I agree, this makes the class cleaner (although an instance is created once when using the `ServiceManager`), fixed in https://github.com/zalando/intellij-swagger/pull/188/commits/255c734bd9ddba0efc19c23568810e79118d449e.

maxim-tschumak commented on examples/extensions-zalando/src/main/java/org/zalando/intellij/swagger/examples/extensions/zalando/validator/zally/ZallyService.java:
```diff
@@ -0,0 +1,7 @@
+import org.zalando.intellij.swagger.examples.extensions.zalando.validator.zally.model.LintingResponse;
+
+public interface ZallyService {
+    LintingResponse validate(final String spec);
```
`validate` is a strong word, let's call it `lint` :)

semonte replied:
I agree, changed `validate` calls to `lint`.

maxim-tschumak commented on examples/extensions-zalando/src/main/java/org/zalando/intellij/swagger/examples/extensions/zalando/validator/zally/http/HttpZallyService.java:
```diff
@@ -0,0 +1,52 @@
+
+    private Map<String, Object> zallyHeaders() {
+        return ImmutableMap.of(
+                "User-Agent", "intellij-swagger-plugin/0.0.1",
```
Is it possible to set the version dynamically? Would be nice to know which plugin versions are used.

semonte replied:
Good idea, fixed in https://github.com/zalando/intellij-swagger/pull/188/commits/13b99f4872a3364b2929caadc01e5b741431e41b

maxim-tschumak commented on examples/extensions-zalando/src/main/java/org/zalando/intellij/swagger/examples/extensions/zalando/validator/zally/http/TokenService.java:
```diff
@@ -0,0 +1,51 @@
+        final String token = new BufferedReader(new
+                InputStreamReader(p.getInputStream())).readLine();
+
+        if (token == null) {
```
In case `ztoken` fails, there will be an error message wich is stored in the `token` variable. It will probably never be a `null`. Should we check the token format? (something like `[\w-.]{256-1024}`)

semonte replied:
The `token` is `null` in case of an error; the error is sent to `p.getErrorStream()`. So I would not add a validation for the format, if it has a value it should be a proper token.

semonte replied:
Path `/usr/loca/bin` was added to the `PATH` in commit https://github.com/zalando/intellij-swagger/pull/188/commits/195a1f7f465f47fcd5ba4b0c4f5152aa4ad0ad90. This will improve the odds that the plugin works out-of-the-box.

maxim-tschumak commented on examples/extensions-zalando/src/main/java/org/zalando/intellij/swagger/examples/extensions/zalando/validator/zally/model/LintingRequest.java:
```diff
@@ -0,0 +1,13 @@
+
+public class LintingRequest {
+
+    @SerializedName("api_definition_string")
```
great!

maxim-tschumak reviewed (approved):
Great!

semonte committed:
Append `/usr/local/bin` for ztoken PATH

Some users might have `ztoken` pointing to `/usr/local/bin`.

semonte committed:
Renamed Zally's `validate` calls to `lint`

semonte committed:
Inject ZallyService once in constructor

semonte committed:
Add plugin version dynamically

In order to track plugin usage, we need to provide
version information to Zally.

semonte commented:
Thanks @maxim-tschumak for the review. I have addressed your comments.

maxim-tschumak commented:
:+1:

semonte committed:
Improved name & description of the plugin
